domain/externalcontroller/modelmigration/export.go

```diff
@@ -0,0 +1,101 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package modelmigration
+
+import (
+	"context"
+
+	"github.com/juju/description/v4"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/crossmodel"
+	"github.com/juju/juju/core/modelmigration"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/domain/externalcontroller"
+	"github.com/juju/juju/domain/externalcontroller/service"
+	"github.com/juju/juju/domain/externalcontroller/state"
+)
+
+// RegisterExport registers the export operations with the given coordinator.
+func RegisterExport(coordinator Coordinator) {
+	coordinator.Add(&exportOperation{})
+}
+
+type ExportService interface {
+	ControllerForModel(ctx context.Context, modelUUID string) (*crossmodel.ControllerInfo, error)
+	ModelsForController(ctx context.Context, controllerUUID string) ([]string, error)
+}
+
+// exportOperation describes a way to execute a migration for
+// exporting external controller s.
+type exportOperation struct {
+	modelmigration.BaseOperation
+
+	service ExportService
+}
+
+func (e exportOperation) Setup(scope modelmigration.Scope) error {
+	// We must not use a watcher during migration, so it's safe to pass a
+	// nil watcher factory.
+	e.service = service.NewService(
+		state.NewState(domain.ConstFactory(scope.ControllerDB())), nil)
+	return nil
+}
+
+// Execute the migration of the external controllers using typed interfaces, to
+// ensure we don't loose any type safety.
+// This export makes use of the remote applications on the source model. Since
+// the source model is stored in mongodb whilst the destination (external)
+// controller is stored in dqlite, we need to be very careful when we fill
+// the dst description.Model argument and make sure that the model is not
+// updated until the export has finished, thus avoiding a race on the
+// remote applications of the model.
+func (e exportOperation) Execute(ctx context.Context, model description.Model) error {
+	// If there are not remote applications, then no external controllers will
+	// be exported. We should understand if that's ever going to be an issue?
+	remoteApplications := model.RemoteApplications()
+
+	// Iterate over the source model UUIDs, to gather up all the related
+	// external controllers. Store them in a map to create a unique set of
+	// source model UUIDs, that way we don't request multiple versions of the
+	// same external controller.
+	sourceModelUUIDs := make(map[string]struct{})
```

Comment:
We have `juju/collections` to do this ergonomically; see `set.NewStrings`.

In any case, we shouldn't be doing this imperatively like this. It's one query to get this data with an `IN` clause for models, then an iteration over the rows to re-normalise into our structs.

Reply:
Indeed, this is a very nice catch and I agree. I rewrote the export in a way we only perform one SQL query and de-normalize into our structs, see 870a42a846fb99d329e45259561d863be3df6e5d